### ISSUE TITLE
Using EF.Property to target included entity property

### ISSUE THREAD

**Issue by Enngage** [aspnet/EntityFrameworkCore#17099]
Hi!

Is it possible to use `EF.Property<>` to target property of included entities? Following attempts:
```c#
 var fail = await Dependencies.EntityServices.FavoriteExerciseService.GetAll()
     .Include(m => m.Exercise)
     .Where(m => EF.Functions.Contains(EF.Property<string>(m, "Exercise.DisplayName"), "\"*press*\""))
     .ToListAsync();

var fail = await Dependencies.EntityServices.FavoriteExerciseService.GetAll()
     .Include(m => m.Exercise)
     .Where(m => EF.Functions.Contains(EF.Property<string>(m, m.Exercise.DisplayName), "\"*press*\""))
     .ToListAsync();
```

throw 

```
The property 'Exercise.DisplayName' on entity type 'FavoriteExercise' could not be found. Ensure that the property exists and has been included in the model.
```

The `DisplayName` on `Exercise` type is properly indexed and this works and returns correct results:

```c#
var ok = await Dependencies.EntityServices.ExerciseService.GetAll()
                .Where(m => EF.Functions.Contains(EF.Property<string>(m, "DisplayName"), "\"*press*\""))
                .ToListAsync();
```

I also tried running following query on SQL which also returns proper results.

```sql
select * from FavoriteExercises join Exercises
on FavoriteExercises.ExerciseId = Exercises.Id
where Contains(Exercises.DisplayName, '"*press*"')
```

Thank you for any help or suggestions.

### Further technical details
EF Core version: 3.0.0-preview5.19227.1
Database Provider: Microsoft.EntityFrameworkCore.SqlServer
Operating system: Win 10
IDE: Visual Studio 2019 preview

**Comment by smitpatel** [comment]
EF.Property can be used to access property or navigation defined in EF Core metadata. But the function does not parse multi-level as you intend. The proper way to write such query would be like this.
```C#
 var fail = await Dependencies.EntityServices.FavoriteExerciseService.GetAll()
     .Include(m => m.Exercise)
     .Where(m => EF.Functions.Contains(EF.Property<string>(EF.Property<Exercise>(m, "Exercise"), "DisplayName"), "\"*press*\""))
     .ToListAsync();
```
The 2nd fail query is not valid since the property or navigation name is not passed as string. Compilation would fail.

**Comment by smitpatel** [comment]
Also you don't need include to access property on a navigation in where predicate. You would only need include if you are projecting the entity and want the navigation to be populated for the materialized entity.

**Comment by Enngage** [comment]
@smitpatel thank you! That indeed works. Though, the problem I have is that `Exercise` type can be anything and it is known only at runtime. Would it be somehow possible to pass the `Type` or `string` instead of using generics?

**Comment by smitpatel** [comment]
The generic constraint on EF.Property is to allow you to compose over while writing query. (e.g. doing `EF.Property<Exercise>(m, "Exercise").Name`). Implementation wise, we just work with the type navigation provides us so what you have written would work.

**Comment by Enngage** [comment]
@smitpatel it seems that this has stopped working in the `preview8` :/ (I'm coming from `preview5` as `preview6` didn't work for me) 

This works ok:

```c#
await Dependencies.EntityServices.FavoriteExerciseService.GetAll()
    .Where(m => EF.Functions.Contains(
       EF.Property<string>(EF.Property<Exercise>(m, "Exercise"), "DisplayName"), "\"*press*\""))
    .ToListAsync();

```

This does not:

```c#
await Dependencies.EntityServices.FavoriteExerciseService.GetAll()
      .Where(m => EF.Functions.Contains(
         EF.Property<string>(EF.Property<IEntity>(m, "Exercise"), "DisplayName"), "\"*press*\""))
      .ToListAsync();
```

Is this behavior intentional? Or maybe there is indeed a way of providing the `type` instead of generics?

**Comment by smitpatel** [comment]
Likely, somewhere a convert to interface node got added which caused client eval. Currently it is not possible for us to bind EF Property to an interface type as we don't have metadata about interface like we have for EntityType.

**Comment by Enngage** [comment]
That seems strange as it just worked :/ Can you think of any workaround? Other than possibly writing the SQL manually as that is not that simple to put it in our code as we are working with IQueryable and have no access DbSet (where sql methods were moved)

**Comment by smitpatel** [comment]
Easy work-around, don't use interfaces. If you use concrete type (which is mapped in EF Core model), it will work.

**Comment by Enngage** [comment]
I do have a `Type`, but I can't use it in the `EF.Property` because that only takes generic argument. Maybe this should be extended to also accept Type as parameter? Entire query is build at runtime so I have to work with string names or Types when necessary.

**Comment by smitpatel** [comment]
I was looking at first post to see query, can you use `m.Exercise` directly no?

**Comment by Enngage** [comment]
Yeah, but the query is just an example. The `Where` condition as in:

```c#
.Where(m => EF.Functions.Contains(EF.Property<string>(EF.Property<IEntity>(m, "Exercise"), "DisplayName"), "\"*press*\""))
```

Is applied dynamically on potentially any DbSet with any property and any navigation property. Imho, it would be awesome to be able to pass type to the `EF.Propery` method as that would likely solve the problem. Though, I'm a bit sad that the previous implementation with interfaces no longer works :( It was indeed producing proper queries on database.

**Comment by ajcvickers** [comment]
Triage: probably need to punt this too
